sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosDataWriteFactory.scala

```diff
@@ -25,27 +27,81 @@ class CosmosDataWriteFactory(userConfig: Map[String, String],
+    val containerDefinition = container.read().block().getProperties
+    val partitionKeyDefinition = containerDefinition.getPartitionKeyDefinition
+
     override def write(internalRow: InternalRow): Unit = {
-      // TODO moderakh: schema is hard coded for now to make end to end TestE2EMain work implement schema inference code
       val objectNode = CosmosRowConverter.fromInternalRowToObjectNode(internalRow, inputSchema)
-      // TODO: moderakh how should we handle absence of id?
-      if (!objectNode.has("id")) {
-        objectNode.put("id", UUID.randomUUID().toString)
+      ensureHasId(objectNode)
+
+      // TODO modreakh investigate if we should also support point write in non-blocking way
+      // TODO moderakh support patch?
+      // TODO moderakh bulkWrite in another PR
+
+      val partitionKeyValue = PartitionKeyHelper.getPartitionKeyPath(objectNode, partitionKeyDefinition)
+
+      if (cosmosWriteConfig.upsertEnabled) {
+        upsertWithRetry(partitionKeyValue, objectNode, cosmosWriteConfig.maxRetryCount + 1)
+      } else {
+        createWithRetry(partitionKeyValue, objectNode, cosmosWriteConfig.maxRetryCount + 1)
+      }
+    }
+
+    private def createWithRetry(partitionKeyValue: PartitionKey,
+                                objectNode: ObjectNode,
+                                remainingAttempts: Int,
+                                exceptionToThrow: Option[Exception] = Option.empty): Unit = {
+      if (remainingAttempts == 0) {
+        assert(exceptionToThrow.isDefined)
+        throw exceptionToThrow.get
+      }
+      try {
+        container.createItem(objectNode, partitionKeyValue, new CosmosItemRequestOptions()).block()
+      } catch {
+        case e: CosmosException if Exceptions.isResourceExistsException(e) => {
+          // DONE
+        }
+        case e: CosmosException => {
+          createWithRetry(partitionKeyValue, objectNode, remainingAttempts - 1, exceptionToThrow)
+        }
+      }
+    }
+
+    private def upsertWithRetry(partitionKeyValue: PartitionKey,
+                                objectNode: ObjectNode,
+                                remainingAttempts: Int,
+                                exceptionToThrow: Option[Exception] = Option.empty): Unit = {
+      if (remainingAttempts == 0) {
+        assert(exceptionToThrow.isDefined)
+        throw exceptionToThrow.get
+      }
+      try {
+        container.upsertItem(objectNode, partitionKeyValue, new CosmosItemRequestOptions()).block()
+      } catch {
+        case e: CosmosException => upsertWithRetry(partitionKeyValue, objectNode, remainingAttempts - 1, Option.apply(e))
+        case e: Exception => throw e // unexpected failure
+      }
+    }
+
+    private def ensureHasId(objectNode: ObjectNode): Unit = {
+      // ensures the id has id otherwise it will autogenerate one
+      if (!objectNode.has(CosmosConstants.CosmosIdFieldName)) {
+        objectNode.put(CosmosConstants.CosmosIdFieldName, UUID.randomUUID().toString)
```

Comment:
The SDK does not have this default behavior, should this be an opt-in config?

Reply:
no more id generation. code removed as per offline discussion.